sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java

```diff
@@ -0,0 +1,747 @@
+ *   or non serializable elements. This {@link DoFn}s are used by default {@link PTransform}s.
+ * </ul>
+ *
+ * <h4>Query the resulting {@link PCollection} containing the sketch.</h4>
+ *
+ * <pre><code>
+ * int k = 100;
+ * int capacity = 100000
+ * {@literal PCollection<StreamSummarySketch<MyClass>>} input = ...;
+ * {@literal PCollection<KV<MyClass, Long>>} output = input
+ *     .apply(MostFrequent.MostFrequentFn.NonSerializableElements
+ *        .create(new MyClassCoder())
+ *        .withCapacity(capacity))
+ *     .apply(ParDo.of(RetrieveTopK.globallyNonSerializable()));
+ * </code></pre>
+ *
+ * <b>Warning: this class is experimental.</b> <br>
+ * Its API is subject to change in future versions of Beam.
+ */
+@Experimental
+public final class MostFrequent {
+
+  /**
+   * Computes and returns the top K most frequent elements in the input {@link PCollection}.
+   *
+   * @param <InputT>    the type of the elements in the input {@link PCollection}
+   */
+  public static <InputT> GlobalSummary<InputT> globally() {
+    return GlobalSummary.<InputT>builder().build();
+  }
+
+  /**
+   * Computes and returns the top K most frequent elements in the input
+   * {@link PCollection} of {@link KV}s.
+   *
+   * @param <K>         the type of the keys mapping the elements
+   * @param <V>         the type of the elements being combined per key
+   */
+  public static <K, V> PerKeySummary<K, V> perKey() {
+    return PerKeySummary.<K, V>builder().build();
+  }
+
+  /** Implementation of {@link #globally()}. */
+  @AutoValue
+  public abstract static class GlobalSummary<InputT>
+          extends PTransform<PCollection<InputT>, PCollection<KV<InputT, Long>>> {
+
+    abstract int capacity();
+    abstract int topK();
+    abstract Builder<InputT> toBuilder();
+
+    static <InputT> Builder<InputT> builder() {
+      return new AutoValue_MostFrequent_GlobalSummary.Builder<InputT>()
+              .setCapacity(10000)
+              .setTopK(100);
+    }
+
+    @AutoValue.Builder
+    abstract static class Builder<InputT> {
+      abstract Builder<InputT> setCapacity(int c);
+      abstract Builder<InputT> setTopK(int k);
+
+      abstract GlobalSummary<InputT> build();
+    }
+
+    /**
+     * Sets the number {@code k} of most frequent element to retrieve.
+     *
+     * @param k the top k most frequent elements to return
+     */
+    public GlobalSummary<InputT> topKElements(int k) {
+      return toBuilder().setTopK(k).build();
+    }
+
+    /**
+     * Sets the capacity {@code c}.
+     *
+     * @param c the capacity of the summary
+     */
+    public GlobalSummary<InputT> withCapacity(int c) {
+      return toBuilder().setCapacity(c).build();
+    }
+
+    @Override
+    public PCollection<KV<InputT, Long>> expand(PCollection<InputT> input) {
+      final Coder<InputT> inputCoder = input.getCoder();
+      Class clazz = inputCoder.getEncodedTypeDescriptor().getRawType();
```

Comment:
This might fail given sufficient type erasure.

Reply:
What do you mean by sufficient type erasure?

Reply:
Disregard this, I'm not sure what I meant anymore.